ciliumidentity resiliency improvement

tommyp1ckles opened the pull request:
Currently, the CiliumIdentity K8s watcher OnDelete handler has a mechanism to recreate wrongly deleted CiliumIdentities.

This is done by checking if:
* If `enableMasterKeyProtection` is enabled (default=true).
* If this agent is the "owning" agent (i.e. the cid originates from this node and the cid is in the agents localStore).

If so, K8s Delete events are handled by recreating the local CID object.  However this is not done resiliently - where if there is a failure at that time with the create request to apiserver then the error is logged and the issue never resolved.

To improve this, this will perform the request attempt in a controller, terminating only if:

* Since the last attempt, the ciliumidentity was not recreated (i.e. still doesn't exist).
* The identity is still locally owned.

This will provide resiliency against issues where the apiserver is temporarily unavailable, the agent has been rate limited, etc.

### Testing

Kind: ensuring ciliumidentities are recreated by deleting the cid and looking for relevant logs: 

```
level=debug msg="OnDelete MasterKeyProtection update succeeded" id=42281 subsys=allocator
```

```release-note
When master key protection is enabled, failed attempts at recreating k8s identity resources will now be retried. 
```

tommyp1ckles committed:
allocator: update ciliumidentity allocator cache to be more resilient.

OnDelete k8s handler will now use a controller to reattempt master key
recreation (i.e. situations where the owning agent sees a delete event
on the k8s cid crd).

This will use the localKeys store as a source of truth of local
ciliumidentities, with sync attempts only happening if the numeric identity
exists in localKeys.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
allocator: only attempt master key recreation on watch delete events.

onDeleteLocked is used when draining the cache to remove all keys.
In this case we want to remove all keys regardless of master keys.

This switches this functionality to make the master key recreation
optional, and only using this behavior when handling OnDelete events.

Other uses of onDeleteLocked should always remove the identity locally.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/allocator: add unit tests for k8s OnDelete handler.

This test should cover:
* OnDelete with master key protection enabled causing resilient sync of
  cid crd.
* OnDelete without master key protection.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles commented on pkg/allocator/cache.go:
```diff
@@ -156,20 +161,57 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+					// then this completes (i.e. the key exists).
+					// Otherwise we will attempt to create the key, this process repeats until
+					// the key is created.
+					if err := a.backend.UpdateKey(ctx, id, value, true); err != nil {
```
For @cilium/kvstore reviewer, there shouldn't be any problems with this when the backend is the etcd-kvstore, however if you are aware of any "leakiness" in this backend api where this might cause issues that would be worth highlighting.

giorio94 replied:
Hmm, in kvstore mode UpdateKey attempts to recreate both the master and slave keys, ensuring that  possible existing ones are never overwritten (i.e., the creation is performed in a transaction). Which also means that the logic is idempotent. 

One possible issue, though, is that if the creation of the master key succeeds, but that of the slave one fails, then it will never be retried again. Indeed, during the next run the `GetByID` check will succeed (IIUC its purpose), and terminate the controller. I wonder if it could make sense to just drop the `GetByID` check, as UpdateKey seems to be idempotent with both backend types.

tommyp1ckles replied:
> One possible issue, though, is that if the creation of the master key succeeds, but that of the slave one fails, then it will never be retried again

Hmm good point, I think the AcquireReference call actually does a getByID lookup in the CRD case anyway.

schlosna commented on pkg/allocator/cache.go:
```diff
@@ -156,20 +161,57 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+}
+
+func syncControllerName(id idpool.ID) string {
+	return fmt.Sprintf("sync-k8s-identity-%s", id)
```
nit: could just use string concat per #19571 though this doesn't seem to be a likely hot path and k8s rpc will dominate any overhead difference

```suggestion
	return "sync-k8s-identity-"+id
```

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -159,17 +164,49 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
-				log.WithError(err).Errorf("OnDelete MasterKeyProtection update for key %q", id)
-			}
+			c.controllers.UpdateController(syncControllerName(id), controller.ControllerParams{
+				Context: context.TODO(),
```
I'd suggest to also specify the `Group` for metrics purposes: https://github.com/cilium/cilium/pull/26850

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -159,17 +164,49 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+					// First, check that the ciliumidentity hasn't already been recreated.
+					// If it has, then we report success and end the controller.
+					// This is necessary because this is handled async off a delete event
+					if cid, err := c.allocator.backend.GetByID(ctx, id); err != nil && cid == nil {
```
Shouldn't it be `err == nil && cid != nil`? IIUC, your intention is to exit if the identity has already been recreated.

tommyp1ckles replied:
Removing this call per last comment

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -159,17 +164,49 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
-			if err != nil {
-				log.WithError(err).Errorf("OnDelete MasterKeyProtection update for key %q", id)
-			}
+			c.controllers.UpdateController(syncControllerName(id), controller.ControllerParams{
```
Should the controllers be stopped when cache.stop is triggered?

tommyp1ckles replied:
Makes sense, I added a stop/drain call on stop following the watcher stopping: https://github.com/cilium/cilium/pull/28912/files#diff-a53523a55dfc263e2ba4590a90e6c21ecfeec344c7cda3011a81f1f993f78f6eR253

This would let any remaining dofuncs complete before closing down.

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -161,17 +161,17 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
 	defer c.mutex.Unlock()
 
-	c.onDeleteLocked(id, key)
+	c.onDeleteLocked(id, key, true)
```
Just wanted to say that the changes in this commit are not strictly required, as master key protection is not enabled for remote cluster allocators (which are the only ones currently using the drain/drainIf methods). But overall make sense to me, making them more future proof.

tommyp1ckles replied:
Yeah I didn't see any actual issue with this but it does make the code a bit harder to reason about.  With this change I was hoping to make this more explicit.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+
+	// 1. Simulate a delete event, where master key protection is enabled
+	// and the identity is owned locally.
+	assert.Nil(t, backend.identities[1234])
```
I guess `require` is more appropriate than `assert` here (and in the other cases), as it immediately terminates the test.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+		return counter.Load() == 3
+	}, time.Millisecond*100, time.Millisecond)
+
+	// 2. Simulate a delete event, where master key protection is enabled
```
s/enabled/disabled

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+	assert.Nil(t, backend.identities[1234])
+	cache := newCache(alloc)
+	cache.OnDelete(1234, TestAllocatorKey("foo"))
+	assert.Eventually(t, func() bool {
```
Nit. EventuallyWithT allows to display better error messages.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+	// and the identity is owned locally.
+	alloc.enableMasterKeyProtection = false
+	cache.OnDelete(1234, TestAllocatorKey("foo"))
+	assert.NotContains(t, backend.identities, 1234, "no master key protection should delete key")
```
s/1234/idpool.ID(1234)/, otherwise the check is always successful due to type mismatch.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+	// 3. Simulate delete event where master key protection is enabled
+	// but the identity is not owned locally.
+	counter.Store(0)
+	alloc.enableMasterKeyProtection = false
```
s/false/true I guess

tommyp1ckles replied:
Test was rewritten to be more reliable, should have these right now but take a look.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,61 @@ func TestObserveAllocatorChanges(t *testing.T) {
+	counter.Store(0)
+	alloc.enableMasterKeyProtection = false
+	cache.OnDelete(1111, TestAllocatorKey("foo"))
+	assert.Equal(t, uint32(0), counter.Load())
```
This check looks unreliable, as it would pass also in the first test (line 561).

tommyp1ckles replied:
👍 Going to rewrite this test a bit and only rely on these for the positive case (i.e. controller runs) but just check the cache state for the negative case

giorio94 reviewed (changes requested):
The overall structure looks good to me. I've left a few comment inline

tommyp1ckles committed:
allocator/allocator_test: remove commented out tests.

These tests have been commented out for over four years, so we should
probably just get rid of them.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -159,17 +163,51 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+					// Otherwise we will attempt to create the key, this process repeats until
+					// the key is created.
+					if err := a.backend.UpdateKey(ctx, id, value, true); err != nil {
+						log.WithError(err).Errorf("OnDelete MasterKeyProtection update for key %q", id)
```
Nit. I guess that this error should no longer be at error level (as now automatically retried), and the id could specified as field rather than as part of the message.

tommyp1ckles replied:
Updated id as field

I think error level still makes sense here, it's a potentially serious issue if recreation never completes and could indicate larger underlying issues.

Although, perhaps moving the success log to Info makes sense to make it clear if the issue has resolved itself.

giorio94 replied:
Makes sense to me. My doubt was mainly whether it should be a warning level, but error sounds reasonable as well (especially because identity recreation is already an unexpected condition to start with).

giorio94 commented on pkg/allocator/cache.go:
```diff
@@ -156,20 +160,55 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+					// For each attempt, check if this ciliumidentity is still a candidate for recreation.
+					// It's possible that since the last iteration that this agent has legitimately deleted
+					// the key, in which case we can stop trying to recreate it.
+					if c.allocator.localKeys.lookupID(id); value == nil {
```
```suggestion
					if value = c.allocator.localKeys.lookupID(id); value == nil {
```

tommyp1ckles replied:
oh nice catch

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -517,6 +524,71 @@ func TestObserveAllocatorChanges(t *testing.T) {
+	alloc.mainCache.OnDelete(1234, TestAllocatorKey("foo"))
+	assert.EventuallyWithT(t, func(c *assert.CollectT) {
+		assertBackendContains(c, 1234, "foo")
+	}, time.Second, time.Millisecond*100)
```
IIUC, this is not testing anything, because the identity is never removed from the backend. Given that the upsertion is already tested through the counter, I guess this could be just removed.

giorio94 reviewed (approved):
Looks good to me. Just a couple more comments inline.
